src/process_user_code.py

```diff
@@ -43,16 +46,39 @@ def run(self):
             read_val = sys.stdin.readline()
             sys.stdin.flush()
             try:
-                new_state = json.loads(read_val)
-                for event in CONSTANTS.EXPECTED_INPUT_EVENTS_CPX:
-                    cpx._Express__state[event] = new_state.get(
-                        event, cpx._Express__state[event]
-                    )
+                new_state_message = json.loads(read_val)
+                device = new_state_message.get(CONSTANTS.ACTIVE_DEVICE_FIELD)
+                new_state = new_state_message.get(CONSTANTS.STATE_FIELD, {})
+
+                if device == CPX:
+                    update_cpx(new_state)
+                elif device == MICROBIT:
+                    update_microbit(new_state)
```

Comment:
Is it possible to get a parent "UserInput" class and inherit classes specific to CPX and microbit? This would make things easier to scale in the future. We wouldn't need to check for the active device in every iteration, etc.

Reply:
It gets a little tricky to instantiate a specific child class based on the device, since the device that the tab is currently on is not known until the first interaction (in this case, the button press) happens. However, the UserInput class is created when the script first starts. If we go for something like that, we would need to pass in an extra argument when we start the python backend based on which tab they are on when they press play, which may take a little time. 
Therefore, I thought that this if..else... was okay given the current structure. What do you think? Sorry for the long read!

Reply:
Technically, we can see which libraries import, but in theory, both libraries could be imported. In terms of keeping track of which device is in use, we're using whichever tab the frontend is on, which is only really kept track of on the frontend.

Reply:
Got it! Do we have some kind of event whenever the button for switching gets pressed? It would be nice if we could at least save some kind of state at UserInput. 

If that ends up being too much work we can prioritize that for later too, I was mostly worried about an overhead/further scaling.

Reply:
We have it so that, if the user presses another tab mid-run, the process stops. This is controlled via the frontend.

Also, I just made a commit that sorta abstracts the device updating away (for better scaling). Now, the updating method is within the microbit/cpx class itself, as per David's comment below! :)

Reply:
When we switch tab, we kill the process. We also tag each message with active_device, so that we can filter out the messages if the device doesn't match.

Reply:
sounds good!